pepper-apis/src/main/java/org/broadinstitute/ddp/service/FileUploadService.java

```diff
@@ -0,0 +1,111 @@
+
+    /**
+     * Service initialization
+     *
+     * @param cfg application configs
+     */
+    public FileUploadService(Config cfg) {
+        boolean ensureDefault = cfg.getBoolean(ConfigFile.REQUIRE_DEFAULT_GCP_CREDENTIALS);
+        GoogleCredentials googleCredentials = GoogleCredentialUtil.initCredentials(ensureDefault);
+        if (googleCredentials != null) {
+            storage = GoogleBucketUtil.getStorage(googleCredentials, cfg.getString(ConfigFile.GOOGLE_PROJECT_ID));
+        } else {
+            throw new IllegalStateException("Need to have backing storage for uploads!");
+        }
+        bucketName = cfg.getString(ConfigFile.FILE_UPLOAD_BUCKET);
+    }
+
+    public String getBucketName() {
+        return bucketName;
+    }
+
+    public Storage getStorage() {
+        return storage;
+    }
+
+    public URL getSignedURLForUpload(Handle handle, String fileUploadGuid, String studyGuid, String activityCode,
+                                     String activityInstanceGuid, String answerGuid, String filename,
+                                     Long fileSize, String mimeType) {
+        String bucketFilename = generateName(studyGuid, activityCode, activityInstanceGuid, answerGuid, fileUploadGuid, filename);
+        BlobInfo blobInfo = BlobInfo.newBuilder(BlobId.of(bucketName,
+                bucketFilename)).build();
+
+        handle.attach(FileUploadDao.class).insertFileUpload(fileUploadGuid, bucketFilename, Instant.now().toEpochMilli(), filename,
+                fileSize, FileUploadStatus.AUTHORIZED);
+
+        Map<String, String> extensionHeaders = new HashMap<>();
+        extensionHeaders.put("Content-Type", mimeType);
+        return storage.signUrl(blobInfo, URL_VALID_TIME_MINUTES, TimeUnit.MINUTES,
+                Storage.SignUrlOption.withV4Signature(),
+                Storage.SignUrlOption.httpMethod(HttpMethod.PUT),
+                Storage.SignUrlOption.withExtHeaders(extensionHeaders));
+    }
+
+    public boolean validateUpload(Handle handle, String guid) {
+        FileUploadDao fileUploadDao = handle.attach(FileUploadDao.class);
+        FileUpload fileUpload = fileUploadDao.getFileUploadByGuid(guid)
+                .orElseThrow(() -> new DaoException("Could not find file upload with guid " + guid));
+        //TODO: validate bucket contents
+        Long bucketFileSize;
+        Long bucketFileCreateTime;
+        Page<Blob> page = storage.list(bucketName, Storage.BlobListOption.prefix(fileUpload.getBucketFileUrl()));
+        Iterator<Blob> iterator = page.getValues().iterator();
+        if (iterator.hasNext()) {
+            Blob blob = iterator.next();
+            bucketFileSize = blob.getSize();
+            bucketFileCreateTime = blob.getCreateTime();
+            if (bucketFileSize != null && bucketFileSize.equals(fileUpload.getFileSize())) {
+                fileUploadDao.setVerified(fileUpload.getFileUploadGuid(), bucketFileCreateTime);
+            }
```

Comment:
You think this method right place to set file object to read only as you were suggesting in our conversation?

Reply:
Yes, as setting read-only should be following verification.

Reply:
If we set readOnly, we won't be able to remove the file after submission